File: Astrolabe.Evaluator/PrintExpr.cs
using System.Text;

namespace Astrolabe.Evaluator;

public static class PrintExpr
{
    // Operator precedence levels (higher number = higher precedence)
    private static class Precedence
    {
        public const int Ternary = 1;    // ? :
        public const int Or = 2;         // or
        public const int And = 3;        // and
        public const int Rel = 4;        // <, >, =, !=, <=, >=
        public const int Plus = 5;       // +, -
        public const int Times = 6;      // *, /, %
        public const int Map = 7;        // .
        public const int Filter = 8;     // [...]
        public const int Prefix = 9;     // unary !, +, -
        public const int Call = 10;      // function calls
    }

    private static int GetPrecedence(EvalExpr expr)
    {
        if (expr is not CallExpr call) return Precedence.Call; // atoms have highest precedence

        return call.Function switch
        {
            "?" => Precedence.Ternary,
            "or" => Precedence.Or,
            "and" => Precedence.And,
            "<" or ">" or "=" or "!=" or "<=" or ">=" => Precedence.Rel,
            "+" or "-" => Precedence.Plus,
            "*" or "/" or "%" => Precedence.Times,
            "." => Precedence.Map,
            "[" => Precedence.Filter,
            "??" => Precedence.Or, // default operator
            _ => Precedence.Call
        };
    }

    private static string PrintWithParens(EvalExpr expr, int parentPrecedence)
    {
        var exprPrecedence = GetPrecedence(expr);
        var needsParens = exprPrecedence < parentPrecedence;
        var printed = expr.Print();
        return needsParens ? $"({printed})" : printed;
    }

    private static string EscapeString(string str)
    {
        var sb = new StringBuilder();
        foreach (var c in str)
        {
            switch (c)
            {
                case '\\': sb.Append("\\\\"); break;
                case '"': sb.Append("\\\""); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                case '\b': sb.Append("\\b"); break;
                case '\f': sb.Append("\\f"); break;
                case '\v': sb.Append("\\v"); break;
                default: sb.Append(c); break;
            }
        }
        return sb.ToString();
    }

    public static string PrintValue(object? value)
    {
        return value switch
        {
            null => "null",
            true => "true",
            false => "false",
            EmptyPath => "",
            DataPath dp => dp.ToPathString(),
            ArrayValue av => $"[{string.Join(", ", av.Values.Select(x => PrintValue(x.Value)))}]",
            ObjectValue ov => PrintObjectValue(ov),
            string s => $"\"{EscapeString(s)}\"",
            _ => $"{value}",
        };
    }

    private static string PrintObjectValue(ObjectValue obj)
    {
        if (!obj.Properties.Any()) return "{}";

        var pairs = obj.Properties.Select(kvp =>
            $"\"{EscapeString(kvp.Key)}\": {kvp.Value.Print()}"
        );
        return $"{{{string.Join(", ", pairs)}}}";
    }

    public static string Print(this EvalExpr expr)
    {
        return expr switch
        {
            ValueExpr v => PrintValue(v.Value),
            VarExpr varExpr => $"${varExpr.Name}",
            LambdaExpr lambdaExpr => $"${lambdaExpr.Variable} => {lambdaExpr.Value.Print()}",
            LetExpr letExpr when !letExpr.Vars.Any() => letExpr.In.Print(),
            LetExpr letExpr =>
                $"let {string.Join(", ", letExpr.Vars.Select(x => $"${x.Item1.Name} := {x.Item2.Print()}"))} in {letExpr.In.Print()}",
            ArrayExpr arrayExpr =>
                $"[{string.Join(", ", arrayExpr.Values.Select(x => x.Print()))}]",
            CallExpr { Function: "[", Args: var a } when a.ToList() is [var first, var t] =>
                $"{PrintWithParens(first, Precedence.Filter)}[{t.Print()}]",
            CallExpr { Function: "?", Args: var a } when a.ToList() is [var ifE, var t, var f] =>
                $"{PrintWithParens(ifE, Precedence.Ternary + 1)} ? {PrintWithParens(t, Precedence.Ternary)} : {PrintWithParens(f, Precedence.Ternary)}",
            CallExpr { Function: "object", Args: var args } when args.Count % 2 == 0 =>
                PrintObjectLiteral(args),
            CallExpr { Function: "string", Args: var args } when args.Count > 1 =>
                PrintTemplateString(args),
            PropertyExpr propExpr => propExpr.Property,
            // Handle unary minus: 0 - expr => -expr
            CallExpr { Function: "-", Args: var args } when args.ToList() is [ValueExpr { Value: 0 }, var operand] =>
                $"-{PrintWithParens(operand, Precedence.Prefix)}",
            // Handle unary NOT: !expr
            CallExpr { Function: "!", Args: var args } when args.Count == 1 =>
                $"!{PrintWithParens(args[0], Precedence.Prefix)}",
            CallExpr callExpr when InfixFunc(callExpr.Function) is { } op && callExpr.Args.ToList() is [var v1, var v2] =>
                PrintBinaryOp(callExpr, v1, v2, op),
            CallExpr callExpr =>
                $"${callExpr.Function}({string.Join(", ", callExpr.Args.Select(x => x.Print()))})",
            _ => expr.ToString()!,
        };
    }

    private static string PrintObjectLiteral(IList<EvalExpr> args)
    {
        if (args.Count == 0) return "{}";

        var pairs = new List<string>();
        for (int i = 0; i < args.Count; i += 2)
        {
            var key = args[i];
            var value = args[i + 1];
            // Print the key as-is to preserve whether it's a string literal or identifier
            pairs.Add($"{key.Print()}: {value.Print()}");
        }
        return $"{{{string.Join(", ", pairs)}}}";
    }

    private static string PrintTemplateString(IList<EvalExpr> args)
    {
        var sb = new StringBuilder("`");
        foreach (var arg in args)
        {
            if (arg is ValueExpr { Value: string s })
            {
                // Escape backticks and backslashes in template strings
                sb.Append(s.Replace("\\", "\\\\").Replace("`", "\\`"));
            }
            else
            {
                sb.Append("{").Append(arg.Print()).Append("}");
            }
        }
        sb.Append("`");
        return sb.ToString();
    }

    private static string PrintBinaryOp(CallExpr callExpr, EvalExpr left, EvalExpr right, string op)
    {
        var precedence = GetPrecedence(callExpr);
        // Left-associative: left operand needs parens if lower precedence,
        // right operand needs parens if lower or equal precedence
        return $"{PrintWithParens(left, precedence)}{op}{PrintWithParens(right, precedence + 1)}";
    }

    public static string? InfixFunc(string func)
    {
        return func switch
        {
            "=" => " = ",
            "<" => " < ",
            "<=" => " <= ",
            ">" => " > ",
            ">=" => " >= ",
            "!=" => " != ",
            "and" => " and ",
            "or" => " or ",
            "??" => " ?? ",
            "+" => " + ",
            "-" => " - ",
            "*" => " * ",
            "/" => " / ",
            "%" => " % ",
            "." => ".",
            _ => null,
        };
    }
}

File: astrolabe-evaluator/src/printExpr.ts
import { CallExpr, EvalExpr, Path, ValueExpr } from "./ast";

// Operator precedence levels (higher number = higher precedence)
const PRECEDENCE = {
  ternary: 1,    // ? :
  or: 2,         // or
  and: 3,        // and
  rel: 4,        // <, >, =, !=, <=, >=
  plus: 5,       // +, -
  times: 6,      // *, /, %
  map: 7,        // .
  filter: 8,     // [...]
  prefix: 9,     // unary !, +, -
  call: 10,      // function calls
};

function getPrecedence(expr: EvalExpr): number {
  if (expr.type === "call") {
    switch (expr.function) {
      case "?": return PRECEDENCE.ternary;
      case "or": return PRECEDENCE.or;
      case "and": return PRECEDENCE.and;
      case "<":
      case ">":
      case "=":
      case "!=":
      case "<=":
      case ">=": return PRECEDENCE.rel;
      case "+":
      case "-": return PRECEDENCE.plus;
      case "*":
      case "/":
      case "%": return PRECEDENCE.times;
      case ".": return PRECEDENCE.map;
      case "[": return PRECEDENCE.filter;
      case "??": return PRECEDENCE.or; // default operator, similar precedence to or
      default: return PRECEDENCE.call;
    }
  }
  return PRECEDENCE.call; // atoms have highest precedence
}

function printWithParens(expr: EvalExpr, parentPrecedence: number): string {
  const exprPrecedence = getPrecedence(expr);
  const needsParens = exprPrecedence < parentPrecedence;
  const printed = printExpr(expr);
  return needsParens ? `(${printed})` : printed;
}

export function printExpr(expr: EvalExpr): string {
  switch (expr.type) {
    case "array":
      return "[" + expr.values.map(printExpr).join(", ") + "]";
    case "lambda":
      return `\$${expr.variable} => ${printExpr(expr.expr)}`;
    case "value":
      return printValue(expr);
    case "property":
      return expr.property;
    case "call":
      return printCall(expr);
    case "var":
      return `\$${expr.variable}`;
    case "let":
      if (expr.variables.length == 0) return printExpr(expr.expr);
      return `let ${expr.variables.map((x) => "$" + x[0].variable + " := " + printExpr(x[1])).join(", ")} in ${printExpr(expr.expr)}`;
  }
}

function escapeString(str: string): string {
  return str
    .replace(/\\/g, "\\\\")
    .replace(/"/g, '\\"')
    .replace(/\n/g, "\\n")
    .replace(/\r/g, "\\r")
    .replace(/\t/g, "\\t")
    .replace(/\x08/g, "\\b")  // backspace character (not word boundary!)
    .replace(/\f/g, "\\f")
    .replace(/\v/g, "\\v");
}

export function printValue({ value }: ValueExpr): string {
  if (value == null) return "null";
  if (typeof value === "boolean") return value.toString();
  if (typeof value === "string") return `"${escapeString(value)}"`;
  if (typeof value === "number") return value.toString();

  // Handle arrays
  if (Array.isArray(value)) {
    return `[${value.map(v => printExpr(v)).join(", ")}]`;
  }

  // Handle objects (Record<string, ValueExpr>)
  if (typeof value === "object") {
    const objValue = value as Record<string, ValueExpr>;
    const entries = Object.entries(objValue);
    if (entries.length === 0) return "{}";

    const pairs = entries.map(([key, val]) =>
      `"${escapeString(key)}": ${printExpr(val)}`
    );
    return `{${pairs.join(", ")}}`;
  }

  // Fallback for any unexpected types
  return String(value);
}

export function printCall(call: CallExpr) {
  const args = call.args;
  const precedence = getPrecedence(call);

  switch (call.function) {
    case ".":
      return `${printWithParens(args[0], precedence)}.${printExpr(args[1])}`;
    case "[":
      return `${printWithParens(args[0], precedence)}[${printExpr(args[1])}]`;
    case "?":
      // Ternary is right-associative, so we need special handling
      return `${printWithParens(args[0], precedence + 1)} ? ${printWithParens(args[1], precedence)} : ${printWithParens(args[2], precedence)}`;
    case "object":
      // Object literals: $object("key1", value1, "key2", value2, ...) => {key1: value1, key2: value2, ...}
      if (args.length % 2 === 0) {
        const pairs: string[] = [];
        for (let i = 0; i < args.length; i += 2) {
          const key = args[i];
          const value = args[i + 1];
          // Print the key as-is to preserve whether it's a string literal or identifier
          pairs.push(`${printExpr(key)}: ${printExpr(value)}`);
        }
        return `{${pairs.join(", ")}}`;
      }
      return `\$${call.function}(${args.map(printExpr).join(", ")})`;
    case "string":
      // Template strings: $string("hello ", expr, " world") => `hello {expr} world`
      // Note: This language uses {expr} for interpolations, not ${expr}
      if (args.length > 1) {
        let result = "`";
        for (const arg of args) {
          if (arg.type === "value" && typeof arg.value === "string") {
            // Escape backticks and backslashes in template strings
            result += arg.value
              .replace(/\\/g, "\\\\")
              .replace(/`/g, "\\`");
          } else {
            result += "{" + printExpr(arg) + "}";
          }
        }
        result += "`";
        return result;
      }
      return `\$${call.function}(${args.map(printExpr).join(", ")})`;
    case "+":
    case "-":
    case "*":
    case "/":
    case "%":
    case "=":
    case "!=":
    case "<":
    case ">":
    case ">=":
    case "<=":
    case "and":
    case "or":
    case "??":
      // Left-associative operators: left operand needs parens if lower precedence,
      // right operand needs parens if lower or equal precedence
      return `${printWithParens(args[0], precedence)} ${call.function} ${printWithParens(args[1], precedence + 1)}`;
    default:
      return `\$${call.function}(${args.map(printExpr).join(", ")})`;
  }
}
export function printPath(path: Path): string {
  if (path.segment != null) {
    const parentStr = printPath(path.parent);
    if (typeof path.segment === "number") {
      // Index access uses bracket notation: [0]
      return `${parentStr}[${path.segment}]`;
    } else {
      // Field access uses dot notation: .field or just field at root
      return parentStr ? `${parentStr}.${path.segment}` : path.segment;
    }
  } else return "";
}
